include/seqan3/core/algorithm/configuration.hpp

```diff
@@ -0,0 +1,1071 @@
+ *
+ * ### Usage
+ *
+ * The entire configuration system is designed to work completely in the background of any algorithm. The implementor
+ * of an algorithm only needs to provide the configurations for the algorithm.
+ * In general the type of a configuration is static (see seqan3::detail::config_element_base) within the context of an algorithm
+ * to select the correct code branches based on a policy-driven design, albeit the stored state might still be a runtime
+ * parameter.
+ * However, in some cases a specific configuration might be known first at runtime but needs to be converted to a static
+ * type for the algorithm in use. To enable a transparent conversion from the runtime parameters to a static type
+ * configuration one can use deferred configs (seqan3::detail::deferred_config_element_base), which are invocable
+ * configurations, that store the runtime parameter and on invocation translate this runtime parameter to a static
+ * configuration. The implementor of an algorithm can achieve this by using the function
+ * seqan3::detail::apply_deferred_configs, which iterates through all configurations and in case
+ * it is a deferred configuration it will invoke the translation function and continue with the modified configuration,
+ * which contains now the static type for the specific configuration.
+ *
+ * ### Pipe Notation
+ *
+ * To enable simple extension of configurations the configuration provides a generic pipe interface for the different
+ * configurations. Thus, a config class can be easily constructed by chaining together different properties.
+ * Consider the following example, where we assume that the config `bar` and the config functor `with_foo` are already
+ * given:
+ *
+ * ```cpp
+ * auto my_cfg = configuration<bar>{} | with_foo;  // my_cfg is now of type configuration<foo, bar>
+ * ```
+ * The `with_foo` functor captures the `configuration<bar>` object and returns a new configuration with the added
+ * `foo`-config.
+
+ * ### Tuple interface
+ *
+ * The configuration exposes a tuple interface. The type_trait functions std::tuple_size and std::tuple_element are
+ * overloaded for this class. To access a specific element one can either use `seqan3::get` or `std::get` to query
+ * the state of the configuration. Considering the example from above one can get the state of
+ * the bar-config either as:
+ *
+ * ```cpp
+ * auto bar_state = get<1>(my_cfg);
+ * ```
+ *
+ * or as:
+ *
+ * ```cpp
+ * auto bar_state = get<bar>(my_cfg);
+ * ```
+ *
+ * In addition every algorithm specific config implementation can overload the get-interface with a enum based
+ * get-interface, such that the correct config is extracted based on some type-independent enum value.
+ * This is necessary for the algorithm implementation, as the algorithm neither knows the order nor the exact type of
+ * all configs (the config type could be a template class).
+ */
+template <config_element_concept ... configs_t>
+class configuration
```

Comment:
Why don't you inherit from std::tuple (or seqan3::pod_tuple)? Then you don't have to reimplement the get interfaces...

Reply:
Also you can just use things like std::tuple_cat to recombine config with new config_elements

Reply:
I had a hell of issues with the get interface  when inheriting it from std::tuple. Also I don’t think that it will reduce much of the complexity but probably the push_front logic can be replaced by tuple_cat. Look into it